### PR TITLE
sparse Adam optimizer

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
@reminisce

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@cjolivier01 could you take a look?

File: src/operator/optimizer_op-inl.h [hunk 1]
```diff
@@ -700,6 +700,146 @@ inline void AdamUpdate(const nnvm::NodeAttrs& attrs,
...
+  template<typename DType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, nnvm::dim_t row_length, DType* out_data,
+    DType* mean_data, DType* var_data, const DType* weight_data, const IType* grad_idx,
+    const DType* grad_data, const DType clip_gradient, const DType beta1, const DType beta2,
+    const DType lr, const DType wd, const DType epsilon, const DType rescale_grad) {
+    using nnvm::dim_t;
+    using namespace mshadow_op;
+    for (dim_t j = 0; j < row_length; j++) {
+      // index in data/mean/var
+      dim_t data_i = grad_idx[i] * row_length + j;
```
Comment (cjolivier01) [inline src/operator/optimizer_op-inl.h]: nit: const?

File: src/operator/optimizer_op-inl.h [hunk 2]
```diff
@@ -700,6 +700,146 @@ inline void AdamUpdate(const nnvm::NodeAttrs& attrs,
...
+      // index in grad
+      dim_t grad_i = i * row_length + j;
```
Comment (cjolivier01) [inline src/operator/optimizer_op-inl.h]: nit: const?

File: src/operator/optimizer_op-inl.h [hunk 3]
```diff
@@ -700,6 +700,146 @@ inline void AdamUpdate(const nnvm::NodeAttrs& attrs,
...
+  if (!grad.storage_initialized() || req == kNullOp) return;
+  CHECK_EQ(req, kWriteInplace) << "kWriteInplace is expected for sparse adam_update";
+  CHECK_GT(weight.shape_.Size(), 0);
+  CHECK_GT(mean.shape_.Size(), 0);
+  CHECK_GT(var.shape_.Size(), 0);
+
+  MSHADOW_REAL_TYPE_SWITCH(weight.type_flag_, DType, {
+    MSHADOW_IDX_TYPE_SWITCH(grad.aux_type(kIdx), IType, {
+      MXNET_ASSIGN_REQ_SWITCH(req, req_type, {
+        DType* weight_data = weight.dptr<DType>();
```
Comment (cjolivier01) [inline src/operator/optimizer_op-inl.h]: Can some of these pointers be const?

File: src/operator/optimizer_op-inl.h (hunk 1 above)
Comment (eric-haibin-lin) [inline src/operator/optimizer_op-inl.h]: good point

**Pull request merged**